cpp/src/join/quadtree_point_to_nearest_polyline.cu

```diff
@@ -146,74 +163,178 @@ struct compute_quadtree_point_to_nearest_polyline {
+          *cudf::column_device_view::create(poly_points_y, stream)});
+
+      auto all_point_indices =
+        thrust::make_transform_iterator(all_point_poly_indices_and_distances,
+                                        [] __device__(auto const &x) { return thrust::get<0>(x); });
+
+      // Allocate vectors for the distances min reduction
+      rmm::device_uvector<uint32_t> point_idxs(point_x.size(), stream);
+      rmm::device_uvector<uint32_t> poly_idxs(point_x.size(), stream);
+      rmm::device_uvector<T> distances(point_x.size(), stream);
+
+      // Fill distances with 0
+      thrust::fill(rmm::exec_policy(stream), distances.begin(), distances.end(), T{0});
+
+      // Reduce the intermediate point/polyline indices to lists of point/polyline index pairs and
+      // distances, selecting the polyline index closest to each point.
+      auto const num_distances =
+        thrust::distance(point_idxs.begin(),
+                         thrust::reduce_by_key(
+                           rmm::exec_policy(stream),
+                           // point indices in
+                           all_point_indices,
+                           all_point_indices + num_point_poly_pairs,
+                           all_point_poly_indices_and_distances,
+                           // point indices out
+                           point_idxs.begin(),
+                           // point/polyline indices and distances out
+                           thrust::make_zip_iterator(
+                             thrust::make_discard_iterator(), poly_idxs.begin(), distances.begin()),
+                           // comparator
+                           thrust::equal_to<uint32_t>(),
+                           // binop to select the point/polyline pair with the smallest distance
+                           [] __device__(auto const &lhs, auto const &rhs) {
+                             T const &d_lhs = thrust::get<2>(lhs);
+                             T const &d_rhs = thrust::get<2>(rhs);
+                             // If lhs distance is 0, choose rhs
+                             if (d_lhs == T{0}) { return rhs; }
+                             // if rhs distance is 0, choose lhs
+                             if (d_rhs == T{0}) { return lhs; }
+                             // If distances to lhs/rhs are the same, choose poly with smallest id
+                             if (d_lhs == d_rhs) {
+                               auto const &i_lhs = thrust::get<1>(lhs);
+                               auto const &i_rhs = thrust::get<1>(rhs);
+                               return i_lhs < i_rhs ? lhs : rhs;
+                             }
+                             // Otherwise choose poly with smallest distance
+                             return d_lhs < d_rhs ? lhs : rhs;
+                           })
+                           .first);
+
+      point_idxs.resize(num_distances, stream);
+      point_idxs.shrink_to_fit(stream);
```

Comment:
shrink_to_fit does a copy... Just something to be aware of in terms of performance. Potential optimization if we ever implement efficient virtual memory based realloc in RMM...  Or, if we find it's a problem and the memory overhead is not, could elide this.

Reply:
Yep, this is a space/time trade. `num_distances` will be smaller than `point_idxs` if all the polylines are in the same spatial quadrant. `point_idxs` is large enough to contain indices for every point in the quadtree, so shrinking here allows us to reuse this memory for the output columns.